### ISSUE TITLE
Android: Gyroscope-based camera rotation

### ISSUE THREAD

**Issue by ClobberXD** [minetest/minetest#8748]
Both PC and Android users use the exact same app(lication) and play together on the same servers. Due to their platform, Android users are at a huge advantage in almost every aspect of Minetest, incl. PvP, building, survival, etc. They did get a small buff in 5.0.0, thanks to the new UI, but it'd be nice to also have gyro-based camera rotation to improve... well, their whole experience. While I'd be very happy to implement this myself, I'm completely in the dark when it comes to MT's Android-related code.

### Some specifics

- Gyro-based camera rotation should be optional, and should default to false.
- Gyro-sensitivity should be customizable.

**Comment by jastevenson303** [comment]
Off-topic: I tried to do split screen on Android but it didn't work.  If it
did would this be able to allow VR with head look?

**Comment by ClobberXD** [comment]
I think VR support for MT is being implemented by @srifqi in #8475. But yeah, this would allow players to turn their head to look around. Fun! :)

**Comment by srifqi** [comment]
> I think VR support for MT is being implemented by @srifqi in #8475.

No, it's different from what my PR does. My PR outputs 360 videos that can be played back using VR or any player that supports 360 video format. This issue requests for gyro-controlled camera, no need for 360 videos.

**Comment by srifqi** [comment]
> I think VR support for MT is being implemented by @srifqi in #8475.

No, this is different from what #8475 does. That PR adds an ability to output 360 videos that can be played back using a VR device or any player that supports 360 video format. This issue requests a gyro-controlled camera, no need for 360 video output.

**Comment by paramat** [comment]
> Gyro-based camera rotation

How does this work? Tilt the device to pan the camera view? Does the device have to be precisely vertical or completely stationary to avoid a pan? Wouldn't this cause irritating and unwanted slow camera pans?

I can't yet see how this helps Android players at all. I don't think they're at a 'huge disadvantage' and don't care if they are at a disadvantage, we're not obligated to compensate for their inferior hardware.

**Comment by MoNTE48** [comment]
> we're not obligated to compensate for their inferior hardware.
Said 'inferior' developer...

**Comment by paramat** [comment]
Please behave, if you continue your hostility towards MT you may eventually be banned from the repo.

**Comment by SmallJoker** [comment]
Please stay on-topic.

**Comment by benrob0329** [comment]
This is also useful for VR headsets like Google Cardboard.

**Comment by paramat** [comment]
Different implementation. For VR the camera points in the direction the smartphone is pointed, that couldn't be used for normal playing.

**Comment by MoNTE48** [comment]
VR is NOT AR
VR = Virtual Reality (Only Virtuality)
AR = Augmented Reality (Camera+Virtuality)

**Comment by ClobberXD** [comment]
> Different implementation. For VR the camera points in the direction the smartphone is pointed, that couldn't be used for normal playing.

Not necessarily. Turning the phone helps in looking around, but swiping would still do most of the work.

**Comment by raymoo** [comment]
For Google Cardboard (mentioned by benrob) specifically, wouldn't you be unable to manipulate the screen?

**Comment by benrob0329** [comment]
Bluetooth controllers?

**Comment by paramat** [comment]
> Different implementation. For VR the camera points in the direction the smartphone is pointed, that couldn't be used for normal playing.

> Not necessarily. Turning the phone helps in looking around, but swiping would still do most of the work.

Yes, necessarily. Playing MT on Android would be bizarre and impractical if players had to swing their whole body around, or swing their arms and head up in the air or down towards the ground, to look in another direction.
However that's exactly what VR requires.

So any gyro controlled camera would have to respond to small limited tilts of the device. 'Rolling' the device left and right would not be pleasant. 'Pitching' up and down not so bad.
This is why i asked how this would work. I can't see any benefit.

**Comment by lhofhansl** [comment]
Why is this low-priority? I get that perhaps nobody is going to work on this (or has the relevant skills), but that does not make it low-priority at all.

The lack of great Android support is a great detriment to Minetest.

I suggest we remove that label again - at least from the issue discussing this.

**Comment by jastevenson303** [comment]
Looking around with VR is awesome.  I mean gyro.  I usually skip over the
negative commentary, where people sometimes downvote an idea just because
they like to read their own words.

**Comment by Desour** [comment]
FYI, SDL has a sensor API.
(If your device isn't supported, I suggest you contribute to SDL.)

I've recently seen someone using gyroscope-based camera rotation on the steam deck: https://www.youtube.com/watch?v=78YZSnNMkMY&t=83s
(I *guess* that device automatically converts the sensor data to joystick movements. @rubenwardy might know more.)

**Comment by medleymedley** [comment]
Nice to see others want this! I've already been playing Luanti with gyroscope aiming on PC (using Steam Input), and every time I play on my phone I intuitively try to tilt it to make fine camera adjustments lol

Optional gyro aiming (with configurable sensitivity) would make playing on touchscreens a ton more intuitive for those who want to use it. Luanti is a game all about pointing to the nodes you want to interact with, which touch screens can make clumsy, it's definitely worth looking into!

**Comment by medleymedley** [comment]
Does anyone know where one could read gyroscope information from Android devices? I know how to get it from one of the java files, but I don't know where in the engine inputs are handed over from java to cpp, despite a lot of digging around

**Comment by Unacceptium** [comment]
i may dont really understand whats discussed here but android precisely outputs gyro data, and also gives a 'game rotation vector' which also uses the built-in compass and gives precise data too. i dont understand why would a dead zone be needed.
from my view the gyro based camera look would be nice for beginner vr headsets like the already mentioned google cardboard or phone-based vr devices, but based on personal experience it would be a (more-) fun (-than practical) way to play some games. but there should be some way that old and new games can prevent it without much code...

https://github.com/user-attachments/assets/65dc7bc4-42da-444b-9032-6b6b29817a33

side note: there are games which gives this as an option, ex.: roblox (may be worst example but i have no better this moment) or Nebulous.io (dont just move the camera but also moves the player)or non-game apps such as 'sky map' (main way to look aroun)
